File: leetcode/1-100/94-binary-tree-inorder-traversal/bench/inorder.py
"""Benchmark workload — Binary Tree Inorder Traversal (LeetCode #94).

Python mirror of bench/inorder.kara. Builds a fresh 63-node balanced tree per iteration
and folds a recursive inorder walk into a rolling hash in visit order. Smaller K
(pure-Python recursion is slow); timed separately, NOT cross-checked. See ../README.md.
"""

import sys


class Node:
    __slots__ = ("val", "left", "right")

    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right


def build(lo, hi, shift):
    if lo > hi:
        return None
    mid = lo + (hi - lo) // 2
    return Node(shift + mid, build(lo, mid - 1, shift), build(mid + 1, hi, shift))


def inorder_fold(node, acc):
    if node is None:
        return
    inorder_fold(node.left, acc)
    acc[0] = (acc[0] * 131 + (node.val + 1)) % 1000000007
    inorder_fold(node.right, acc)


def main():
    total = 40000
    modulus = 1000000007
    size = 63
    total_sum = 0
    for k in range(total):
        shift = k % 1000
        root = build(0, size - 1, shift)
        acc = [0]
        inorder_fold(root, acc)
        total_sum = (total_sum * 131 + acc[0]) % modulus
    print(total_sum)


if __name__ == "__main__":
    sys.setrecursionlimit(10000)
    main()

File: leetcode/1-100/94-binary-tree-inorder-traversal/bench/inorder.rs
// Benchmark workload — Binary Tree Inorder Traversal (LeetCode #94).
// Rust mirror of bench/inorder.kara. Each of K=320,000 iterations builds a fresh 63-node
// balanced tree (Box per node), folds a recursive inorder walk into a rolling hash in
// visit order; the Box tree drops at end of iteration — the Rust analogue of Kāra's
// per-iteration RC build/traverse/drop. The tree is read-only during traversal (no
// aliasing, no mutation), so single-owner Box is the honest node — same choice as the
// sibling tree kata #98. Compiled with `rustc -O`. See ../README.md § Benchmarks.

struct Node {
    val: i64,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn build(lo: i64, hi: i64, shift: i64) -> Option<Box<Node>> {
    if lo > hi {
        return None;
    }
    let mid = lo + (hi - lo) / 2;
    Some(Box::new(Node {
        val: shift + mid,
        left: build(lo, mid - 1, shift),
        right: build(mid + 1, hi, shift),
    }))
}

fn inorder_fold(node: &Option<Box<Node>>, acc: &mut i64) {
    if let Some(n) = node {
        inorder_fold(&n.left, acc);
        *acc = (*acc * 131 + (n.val + 1)) % 1_000_000_007;
        inorder_fold(&n.right, acc);
    }
}

fn main() {
    const TOTAL: i64 = 320_000;
    const MODULUS: i64 = 1_000_000_007;
    const SIZE: i64 = 63;
    let mut sum = 0i64;
    for k in 0..TOTAL {
        let shift = k % 1000;
        let root = build(0, SIZE - 1, shift);
        let mut acc = 0i64;
        inorder_fold(&root, &mut acc);
        sum = (sum * 131 + acc) % MODULUS;
    }
    println!("{}", sum);
}
